### PR TITLE
Finish up AnyCalendar formatting features work

### LINKED ISSUE #1673: AnyCalendar support
We should support `AnyCalendar`, which is aware of all `icu_calendar` calendars. This will just be an enum wrapper around specific ICU calendars.

This includes parsing `u-ca` strings

### PR DESCRIPTION
**Pull request by Manishearth**
Fixes #1673

This can't be tested until https://github.com/unicode-org/icu4x/issues/2145, but the feature work is done.

This adds:
 - Proper conversion for AnyDTF
 - Support for ZonedAnyDTF

This still does not do the Grand Renaming that renames AnyDTF to DTF and DTF to TypedDTF, which I consider polish (also I'm not 100% convinced we should do this, since the AnyDTF machinery is heavier).

### CODE REVIEW

File: components/datetime/src/any/datetime.rs [hunk 1]
```diff
@@ -146,7 +144,7 @@ impl AnyDateTimeFormat {
     /// preferable to use a provider that implements `ResourceProvider<D>` for all `D`, and ensure data is loaded as
     /// appropriate. The [`Self::try_new_with_buffer_provider()`], [`Self::try_new_with_any_provider()`] constructors
     /// may also be used if compile stability is desired.
-    ///
+    ///inutes().unwrap_or_def
```
Comment (sffc) [inline components/datetime/src/any/datetime.rs:147]: typo

File: components/datetime/src/any/datetime.rs [hunk 2]
```diff
@@ -203,34 +220,31 @@ impl AnyDateTimeFormat {
...
     /// that contains all information necessary to display a formatted date and operate on it.
-    ///
-    /// This function will fail if the date passed in uses a different calendar than that of the
-    /// AnyCalendar. Please convert dates before passing them in if necessary.
     #[inline]
-    pub fn format<'l, T>(
-        &'l self,
-        value: &'l T,
-    ) -> Result<FormattedDateTime<'l>, DateTimeFormatError>
+    pub fn format<'l, T>(&'l self, value: &T) -> FormattedDateTime<'l>
```
Comment (sffc) [inline components/datetime/src/any/datetime.rs:224]: Why did you change this behavior from being fallible to auto-converting?

File: components/datetime/src/any/zoned_datetime.rs [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+        zone_provider: &ZP,
+        plural_provider: &PP,
+        decimal_provider: &DEP,
+        calendar_provider: &CEP,
+        date_time_format_options: &DateTimeFormatOptions,
+        time_zone_format_options: &TimeZoneFormatOptions,
+    ) -> Result<Self, DateTimeFormatError>
+    where
+        L: Into<Locale>,
+        DP: ResourceProvider<DateSymbolsV1Marker>
```
Comment (sffc) [inline components/datetime/src/any/zoned_datetime.rs:136]: Thought: With `try_new_unstable` we should eventually be able to collapse the providers into one and use a forking provider to delegate between them at the call site. @robertbastian tried getting that to work in ForkByKey but it's a bit tricky. It may require a macro to generate the match statement. Not sure if we want to make this change now or later.

File: components/datetime/src/any/datetime.rs (hunk 2 above)
Comment (Manishearth) [reply components/datetime/src/any/datetime.rs:224]: That's what we had discussed, no?

File: components/datetime/src/any/zoned_datetime.rs (hunk 3 above)
Comment (Manishearth) [reply components/datetime/src/any/zoned_datetime.rs:136]: Yeah. I didn't want to change the API too much in this change.

File: components/datetime/src/any/datetime.rs (hunk 2 above)
Comment (Manishearth) [reply components/datetime/src/any/datetime.rs:224]: In the previous PR you had asked for that error mode to be removed and I had agreed to do that in a followup.

Comment (Manishearth) [reply components/datetime/src/any/datetime.rs:224]: Oh, that was for a different error. Either way, I was under the impression we wanted this to be autoconverting. If not, I can revert.

**Review by sffc — APPROVED** [review]
